61484 prime notoptions within delete option

pbearne opened the pull request:
Update delete_option to modify notoptions cache

This commit introduces updates to the delete_option and delete_network_option functions so that they modify the notoptions cache. Corresponding unit tests have been added to ensure the behavior is as expected. This ensures the query count doesn't increase when calling get_option or get_network_option after an option has been deleted.

Trac ticket: https://core.trac.wordpress.org/ticket/61484

Commit:
Update delete_option to modify notoptions cache

This commit introduces updates to the delete_option and delete_network_option functions so that they modify the notoptions cache. Corresponding unit tests have been added to ensure the behavior is as expected. This ensures the query count doesn't increase when calling get_option or get_network_option after an option has been deleted.

Commit:
Update delete_option to modify notoptions cache

This commit introduces updates to the delete_option and delete_network_option functions so that they modify the notoptions cache. Corresponding unit tests have been added to ensure the behavior is as expected. This ensures the query count doesn't increase when calling get_option or get_network_option after an option has been deleted.

peterwilsoncc commented on tests/phpunit/tests/option/option.php:
```diff
@@ -478,4 +478,23 @@ public function test_update_option_with_autoload_change_yes_to_no() {
+
+		$before = get_num_queries();
+		get_option( 'foo' );
+		$after = get_num_queries();
+
+		$this->assertSame( $after, $before, 'The notoptions cache was not hit.' );
```
This is to make it clearer that no queries are expected. If it fails it will produce the error that 1 does not match 0 rather than something along the lines of 1234 does not match 1233.

```suggestion
		$queries = get_num_queries() - $before;

		$this->assertSame( 0, $queries, 'get_option should not make any database queries.' );
```

peterwilsoncc commented on tests/phpunit/tests/option/option.php:
```diff
@@ -478,4 +478,23 @@ public function test_update_option_with_autoload_change_yes_to_no() {
+	public function test_check_delete_option_updates_notoptions() {
+		add_option( 'foo', 'value1' );
+
+		delete_option( 'foo' );
```
```suggestion
		delete_option( 'foo' );
		
		$notoptions = wp_cache_get( 'notoptions', 'options' );
		$this->assertIsArray( $notoptions, 'The notoptions cache is expected to be an array.' );
		$this->assertTrue( $notoptions['foo'], 'The deleted options is expected to be in notoptions.' );
```

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -1217,6 +1217,16 @@ function delete_option( $option ) {
+
+		$notoptions = wp_cache_get( 'notoptions', 'options' );
+
+		if ( ! is_array( $notoptions ) ) {
+			$notoptions = array();
+		} else{
+			$notoptions[ $option ] = true;
+		}
```
```suggestion
		if ( ! is_array( $notoptions ) ) {
			$notoptions = array();
		}
		$notoptions[ $option ] = true;
```

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -2284,6 +2294,16 @@ function delete_network_option( $network_id, $option ) {
 		 */
 		do_action( 'delete_site_option', $option, $network_id );
 
+		$notoptions_key = "$network_id:notoptions";
+		$notoptions = wp_cache_get( $notoptions_key, 'site-options' ); // Yes, again... we need it to be fresh.
```
CS: Alignment. Cache is not queried earlier in the function.

```suggestion
		$notoptions_key = "$network_id:notoptions";
		$notoptions     = wp_cache_get( $notoptions_key, 'site-options' );
```

peterwilsoncc reviewed (commented):
Added some notes based on a read through of the code.

I've added the notes against the options code but the same applies to the network options code.

pbearne committed:
Update tests/phpunit/tests/option/option.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/option/option.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/option.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/option.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

peterwilsoncc commented:
@pbearne Are you able to make the same changes to the network options code?

mukeshpanchal27 commented on tests/phpunit/tests/option/option.php:
```diff
@@ -478,4 +478,27 @@ public function test_update_option_with_autoload_change_yes_to_no() {
+		add_option( 'foo', 'value1' );
+
+		delete_option( 'foo' );
+		
```
```suggestion

```

Fix PHPCS

pbearne committed:
Update tests/phpunit/tests/option/option.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne commented:
> @pbearne Are you able to make the same changes to the network options code?

added to  delete_network_option to we need it elsewhere?

peterwilsoncc commented:
I mean the changes I suggested in the original review. I made them against the options functions and tests but they're also needed for the network options.

Commit:
Merge branch 'refs/heads/trunk' into 61484-prime-notoptions-within-delete_option

Commit:
Add assertions to networkOption test

The changes include additional assertions to the networkOption unit test in the PHPUnit suite. After a network option is deleted, this update verifies if the option is in the 'notoptions' cache as expected and ensures that the 'notoptions' cache is an array, therefore enhancing the test's robustness and accuracy.

pbearne commented:
@peterwilsoncc made the changes

Commit:
Add 'ms-required' group tag to networkOption test

The test for the delete_option function in the networkOption.php unit test file has been marked with the 'ms-required' group tag. This will allow the test to be grouped and possibly run separately from those not flagged as ms-required.

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -2284,6 +2293,16 @@ function delete_network_option( $network_id, $option ) {
+		$notoptions_key = "$network_id:notoptions";
+		$notoptions     = wp_cache_get( $notoptions_key, 'site-options' );
+
+		if ( ! is_array( $notoptions ) ) {
+			$notoptions = array();
+		} else {
+			$notoptions[ $option ] = true;
+		}
```
```suggestion
		if ( ! is_array( $notoptions ) ) {
			$notoptions = array();
		}
		$notoptions[ $option ] = true;
```

peterwilsoncc commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -56,6 +56,30 @@ public function test_delete_network_option_on_only_one_network() {
 	}
 
+	/**
+	 * Tests that calling delete_option() update nooptions when option deleted.
```
```suggestion
	 * Tests that calling delete_network_option() updates nooptions when option deleted.
```

peterwilsoncc commented on tests/phpunit/tests/option/option.php:
```diff
@@ -478,4 +478,26 @@ public function test_update_option_with_autoload_change_yes_to_no() {
 	}
+
+	/**
+	 * Tests that calling delete_option() update notoptions when option deleted.
```
```suggestion
	 * Tests that calling delete_option() updates notoptions when option deleted.
```

peterwilsoncc commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -56,6 +56,30 @@ public function test_delete_network_option_on_only_one_network() {
+		add_network_option( 1, 'foo', 'value1' );
+
+		delete_network_option( 1, 'foo' );
+		$notoptions = wp_cache_get( '1:notoptions', 'site-options' );
```
```suggestion
		$cache_key   = is_multisite() ? '1:notoptions' : 'notoptions';
		$cache_group = is_multisite() ? 'site-options' : 'options';
		$notoptions  = wp_cache_get( $cache_key, $cache_group );
```

peterwilsoncc commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -56,6 +56,30 @@ public function test_delete_network_option_on_only_one_network() {
+
+		$before = get_num_queries();
+		get_network_option( 1, 'foo' );
+		$after = get_num_queries();
```
```suggestion
		$queries = get_num_queries() - $before;
```

peterwilsoncc commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -56,6 +56,30 @@ public function test_delete_network_option_on_only_one_network() {
+		get_network_option( 1, 'foo' );
+		$after = get_num_queries();
+
+		$this->assertSame( $after, $before, 'The notoptions cache was not hit.' );
```
```suggestion
		$this->assertSame( 0, $queries, 'get_network_option should not make any database queries.' );
```

peterwilsoncc commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -56,6 +56,30 @@ public function test_delete_network_option_on_only_one_network() {
+	/**
+	 * Tests that calling delete_option() update nooptions when option deleted.
+	 *
+	 * @group ms-required
+	 *
+	 * @ticket 61484
```
```suggestion
	 * @ticket 61484
```

With the cache group suggestion, this can run on both single and multisite installs,

peterwilsoncc reviewed (commented):
Added a few more notes inline.

pbearne committed:
Update src/wp-includes/option.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/option/option.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

peterwilsoncc committed:
Further notoptions tests for adding and updating.

peterwilsoncc reviewed (approved):
Thanks Paul, this looks good to me.

I took the liberty of pushing some further tests in b2bf00c64ce734239c0585e57abf87ec00ba7dbd to ensure that the notoptions caches are cleared with the add and updating functions.

They're probably not strictly related to this test but if the notoptions cache is not cleared properly, it would cause problems for the new code here.

peterwilsoncc committed:
CS: Alignment.

peterwilsoncc commented:
Merged in [r58782](https://core.trac.wordpress.org/changeset/58782) / https://github.com/WordPress/wordpress-develop/commit/fc33ab699972a6ff1893707f852abd4e15f4dc65

anomiex commented:
FYI: https://core.trac.wordpress.org/ticket/61730
